File: client/ios/MoltChat/MChatRuntime.swift
import Foundation
import Combine

/// MChat 运行时，对标 Android MChatRuntime
@MainActor
final class MChatRuntime: ObservableObject {
    let prefs = SecurePrefs()
    private let historyCache = ChatHistoryCache()
    
    @Published private(set) var connectionState: MChatConnectionState = .disconnected
    @Published private(set) var isConnected = false
    @Published var chatEmployees: [EmployeeEntry] = []
    @Published var chatMessagesByPeer: [String: [ChatMessage]] = [:]
    @Published var chatCurrentPeerId: String?
    @Published var chatErrorText: String?
    @Published var chatSessions: [ChatSessionEntry] = []
    
    private var connection: MChatConnection?
    private var chatController: MChatChatController?
    private var cancellables = Set<AnyCancellable>()
    private var connectionStateCancellable: AnyCancellable?
    
    var mqttBrokerUrl: String {
        get { prefs.mqttBrokerUrl }
        set { prefs.setMqttBrokerUrl(newValue) }
    }
    var mqttUsername: String {
        get { prefs.mqttUsername }
        set { prefs.setMqttUsername(newValue) }
    }
    var mqttPassword: String {
        get { prefs.mqttPassword }
        set { prefs.setMqttPassword(newValue) }
    }
    var mchatEmployeeId: String {
        get { prefs.mchatEmployeeId }
        set { prefs.setMchatEmployeeId(newValue) }
    }
    
    init() {
        prefs.objectWillChange
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &cancellables)
    }
    
    func connect() {
        let url = prefs.mqttBrokerUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        let username = prefs.mqttUsername.trimmingCharacters(in: .whitespacesAndNewlines)
        let password = prefs.mqttPassword
        let employeeId = prefs.mchatEmployeeId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !url.isEmpty, !username.isEmpty, !employeeId.isEmpty else {
            connectionState = .error("请填写 Broker 地址、用户名与员工 ID")
            return
        }
        let (host, port, useTls) = parseBrokerUrl(url)
        guard let host = host, let port = port else {
            connectionState = .error("Broker 地址格式错误，应为 tcp://host:port 或 ssl://host:port")
            return
        }
        disconnect()
        connectionState = .connecting
        let conn = MChatConnection(
            brokerHost: host,
            brokerPort: port,
            useTls: useTls,
            username: username,
            password: password,
            employeeId: employeeId
        )
        Task { @MainActor in
            let cached = await historyCache.loadAll(myEmployeeId: employeeId)
            let controller = MChatChatController(
                connection: conn,
                historyCache: historyCache,
                initialMessagesByPeer: cached.isEmpty ? nil : cached
            )
            conn.setInboxCallback { [weak controller] payload in
                Task { @MainActor in
                    controller?.handleInboxMessage(payload)
                }
            }
            controller.onEmployeesChanged = { [weak self] in self?.chatEmployees = $0 }
            controller.onMessagesChanged = { [weak self] in self?.chatMessagesByPeer = $0 }
            controller.onCurrentPeerChanged = { [weak self] in self?.chatCurrentPeerId = $0 }
            controller.onErrorChanged = { [weak self] in self?.chatErrorText = $0 }
            controller.onSessionsChanged = { [weak self] in self?.chatSessions = $0 }
            connectionStateCancellable?.cancel()
            connectionStateCancellable = conn.stateSubject
                .receive(on: DispatchQueue.main)
                .sink { [weak self] state in
                    self?.connectionState = state
                    self?.isConnected = (state == .connected)
                    if case .connected = state {
                        controller.selectPeer(nil)
                        controller.loadEmployees()
                    }
                }
            connection = conn
            chatController = controller
            conn.connect()
        }
    }
    
    func disconnect() {
        connection?.disconnect()
        connection = nil
        chatController = nil
        connectionStateCancellable?.cancel()
        connectionStateCancellable = nil
    }
    
    func loadEmployees() { chatController?.loadEmployees() }
    func selectPeer(_ employeeId: String?) { chatController?.selectPeer(employeeId) }
    func sendMessage(peerEmployeeId: String, text: String) {
        chatController?.sendMessage(peerEmployeeId: peerEmployeeId, text: text)
    }
    func currentChatMessages() -> [ChatMessage] { chatController?.currentMessages() ?? [] }
    
    private func parseBrokerUrl(_ url: String) -> (host: String?, port: Int?, useTls: Bool) {
        let trimmed = url.trimmingCharacters(in: .whitespacesAndNewlines)
        let useTls = trimmed.lowercased().hasPrefix("ssl://")
        let prefix = useTls ? "ssl://" : "tcp://"
        guard trimmed.lowercased().hasPrefix(prefix) else { return (nil, nil, false) }
        let rest = String(trimmed.dropFirst(prefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
        guard let colon = rest.firstIndex(of: ":") else { return (rest.isEmpty ? nil : rest, nil, useTls) }
        let host = String(rest[..<colon]).trimmingCharacters(in: .whitespacesAndNewlines)
        let portStr = String(rest[rest.index(after: colon)...]).prefix(5).filter { $0.isNumber }
        let port = Int(portStr).map { min(max($0, 1), 65535) }
        return (host.isEmpty ? nil : host, port, useTls)
    }
}

File: client/android/app/src/main/java/com/clawdbot/android/MChatRuntime.kt
package com.clawdbot.android

import android.content.Context
import com.clawdbot.android.mchat.MChatChatController
import com.clawdbot.android.mchat.MChatConnection
import com.clawdbot.android.mchat.MChatConnectionState
import com.clawdbot.android.mchat.ChatHistoryCache
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * MChat 运行时：仅包含连接配置、MQTT 连接与聊天（员工列表、单聊收发）。
 */
class MChatRuntime(context: Context) {
  private val appContext = context.applicationContext
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

  val prefs = SecurePrefs(appContext)

  private var connection: MChatConnection? = null
  private var chatController: MChatChatController? = null

  private val _connectionState = MutableStateFlow<MChatConnectionState>(MChatConnectionState.Disconnected)
  val connectionState: StateFlow<MChatConnectionState> = _connectionState.asStateFlow()

  private val _isConnected = MutableStateFlow(false)
  val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

  val mqttBrokerUrl: StateFlow<String> = prefs.mqttBrokerUrl
  val mqttUsername: StateFlow<String> = prefs.mqttUsername
  val mqttPassword: StateFlow<String> = prefs.mqttPassword
  val mchatEmployeeId: StateFlow<String> = prefs.mchatEmployeeId

  private val _chatEmployees = MutableStateFlow<List<MChatChatController.EmployeeEntry>>(emptyList())
  val chatEmployees: StateFlow<List<MChatChatController.EmployeeEntry>> = _chatEmployees.asStateFlow()
  private val _chatMessagesByPeer = MutableStateFlow<Map<String, List<com.clawdbot.android.chat.ChatMessage>>>(emptyMap())
  val chatMessagesByPeer: StateFlow<Map<String, List<com.clawdbot.android.chat.ChatMessage>>> = _chatMessagesByPeer.asStateFlow()
  private val _chatCurrentPeerId = MutableStateFlow<String?>(null)
  val chatCurrentPeerId: StateFlow<String?> = _chatCurrentPeerId.asStateFlow()
  private val _chatErrorText = MutableStateFlow<String?>(null)
  val chatErrorText: StateFlow<String?> = _chatErrorText.asStateFlow()
  private val _chatSessions = MutableStateFlow<List<com.clawdbot.android.chat.ChatSessionEntry>>(emptyList())
  val chatSessions: StateFlow<List<com.clawdbot.android.chat.ChatSessionEntry>> = _chatSessions.asStateFlow()

  fun setMqttBrokerUrl(value: String) = prefs.setMqttBrokerUrl(value)
  fun setMqttUsername(value: String) = prefs.setMqttUsername(value)
  fun setMqttPassword(value: String) = prefs.setMqttPassword(value)
  fun setMcHatEmployeeId(value: String) = prefs.setMcHatEmployeeId(value)

  private val historyCache = ChatHistoryCache(appContext)

  init {
    scope.launch {
      if (hasRequiredConfig()) connect()
    }
  }

  private fun hasRequiredConfig(): Boolean {
    val url = prefs.mqttBrokerUrl.value.trim()
    val username = prefs.mqttUsername.value.trim()
    val employeeId = prefs.mchatEmployeeId.value.trim()
    if (url.isEmpty() || username.isEmpty() || employeeId.isEmpty()) return false
    val (host, port, _) = parseBrokerUrl(url)
    return host != null && port != null
  }

  fun connect() {
    val url = prefs.mqttBrokerUrl.value.trim()
    val username = prefs.mqttUsername.value.trim()
    val password = prefs.mqttPassword.value
    val employeeId = prefs.mchatEmployeeId.value.trim()
    if (url.isEmpty() || username.isEmpty() || employeeId.isEmpty()) {
      _connectionState.value = MChatConnectionState.Error("请填写 Broker 地址、用户名与员工 ID")
      return
    }
    val (host, port, useTls) = parseBrokerUrl(url)
    if (host == null || port == null) {
      _connectionState.value = MChatConnectionState.Error("Broker 地址格式错误，应为 tcp://host:port 或 ssl://host:port")
      return
    }
    disconnect()
    _connectionState.value = MChatConnectionState.Connecting
    scope.launch {
      val cached = withContext(Dispatchers.IO) { historyCache.loadAll(employeeId) }
      val conn = MChatConnection(
        scope = scope,
        brokerHost = host,
        brokerPort = port,
        useTls = useTls,
        username = username,
        password = password,
        employeeId = employeeId,
      )
      val controller = MChatChatController(
        scope = scope,
        connection = conn,
        historyCache = historyCache,
        initialMessagesByPeer = cached,
      )
      conn.setInboxCallback { controller.handleInboxMessage(it) }
      connection = conn
      chatController = controller
      launch {
        conn.state.collect { state ->
          _connectionState.value = state
          _isConnected.value = state is MChatConnectionState.Connected
          if (state is MChatConnectionState.Connected) {
            controller.selectPeer(null)
            controller.loadEmployees()
          }
        }
      }
      launch { controller.employees.collect { _chatEmployees.value = it } }
      launch { controller.messagesByPeer.collect { _chatMessagesByPeer.value = it } }
      launch { controller.currentPeerId.collect { _chatCurrentPeerId.value = it } }
      launch { controller.errorText.collect { _chatErrorText.value = it } }
      launch { controller.sessions.collect { _chatSessions.value = it } }
      conn.connect()
    }
  }

  fun disconnect() {
    connection?.disconnect()
    connection = null
    chatController = null
  }

  fun loadEmployees() = chatController?.loadEmployees()
  fun selectPeer(employeeId: String?) = chatController?.selectPeer(employeeId)
  fun sendMessage(peerEmployeeId: String, text: String) = chatController?.sendMessage(peerEmployeeId, text)
  fun getEmployee(employeeId: String, onResult: (MChatChatController.EmployeeEntry?) -> Unit) =
    chatController?.getEmployee(employeeId, onResult)

  fun currentChatMessages(): List<com.clawdbot.android.chat.ChatMessage> =
    chatController?.currentMessages() ?: emptyList()

  private fun parseBrokerUrl(url: String): Triple<String?, Int?, Boolean> {
    val trimmed = url.trim()
    val useTls = trimmed.startsWith("ssl://", ignoreCase = true)
    val prefix = if (useTls) "ssl://" else "tcp://"
    if (!trimmed.lowercase().startsWith(prefix)) return Triple(null, null, false)
    val rest = trimmed.substring(prefix.length).trim()
    val colon = rest.indexOf(':')
    if (colon <= 0) return Triple(rest.ifEmpty { null }, null, useTls)
    val host = rest.substring(0, colon).trim()
    val portStr = rest.substring(colon + 1).trim().takeWhile { it.isDigit() }
    val port = portStr.toIntOrNull()?.coerceIn(1, 65535)
    return Triple(host.ifEmpty { null }, port, useTls)
  }
}
